### PR TITLE
Optimize M2L for GPU

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

**Comment by isuruf** [comment]
@inducer, this is ready for review

File: sumpy/e2e.py [hunk 1]
```diff
@@ -515,7 +516,9 @@ def get_kernel(self, result_dtype):
     def get_optimized_kernel(self, result_dtype):
         knl = self.get_kernel(result_dtype)
         # FIXME
```
Comment (inducer) [inline sumpy/e2e.py:518]:
```suggestion
```

You're adding the missing transform, after all.

File: sumpy/e2e.py [hunk 2]
```diff
@@ -515,7 +516,9 @@ def get_kernel(self, result_dtype):
...
-        knl = lp.split_iname(knl, "itgt_box", 16, outer_tag="g.0")
+        knl = lp.tag_inames(knl, dict(itgt_box="g.0"))
```
Comment (inducer) [inline sumpy/e2e.py:519]: How come you decided to leave this outside the M2L transform? Surely it interacts with the code there?

File: sumpy/expansion/m2l.py [hunk 3]
```diff
@@ -740,8 +743,25 @@ def postprocess_local_exprs(self, tgt_expansion, src_expansion, m2l_result,
         return super().postprocess_local_exprs(tgt_expansion,
             src_expansion, m2l_result, src_rscale, tgt_rscale, sac)
 
-# }}} VolumeTaylorM2LWithFFT
+    def optimize_loopy_kernel(self, knl, tgt_expansion, src_expansion):
+        # Transform the kernel so that icoeff_tgt and its duplicates
+        # become the outermost iname
```
Comment (inducer) [inline sumpy/expansion/m2l.py:748]: Should this be gated to be a GPU-only transform, or is this also helpful on the CPU?

**Review by inducer — COMMENTED** [review]
Thanks! Just a few nits below.

File: sumpy/e2e.py (hunk 2 above)
Comment (isuruf) [reply sumpy/e2e.py]: Fixed in 2746b9a01e88e749a37308dd0386c6c8ae9ed613

File: sumpy/e2e.py (hunk 1 above)
Comment (isuruf) [reply sumpy/e2e.py:518]: Fixed in 2746b9a01e88e749a37308dd0386c6c8ae9ed613

File: sumpy/expansion/m2l.py [hunk 4]
```diff
@@ -738,6 +741,43 @@ def postprocess_local_exprs(self, tgt_expansion, src_expansion, m2l_result,
...
+    def optimize_loopy_kernel(self, knl, tgt_expansion, src_expansion):
+        from loopy.translation_unit import for_each_kernel
+        from loopy.match import parse_stack_match
+        from pymbolic import var
+        from pymbolic.mapper.substitutor import make_subst_func
+        from loopy.symbolic import (RuleAwareSubstitutionMapper,
+            SubstitutionRuleMappingContext)
+
+        # Transform the kernel so that icoeff_tgt and its duplicates
+        # become the top most iname
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/m2l.py]: Changed to `outermost`.

**Comment by inducer** [comment]
LGTM, thanks!

**Pull request merged**